source/opt/passes.h

```diff
@@ -74,6 +80,165 @@ class EliminateDeadConstantPass : public Pass {
+  // true if the spec constant is folded to normal constants. New instructions
+  // will be inserted before the OpSpecConstantOp instruction pointed by the
+  // instruction iterator. The instruction iterator, which is passed by
+  // pointer, will still point to the original OpSpecConstantOp instruction. If
+  // folding is done successfully, the original OpSpecConstantOp instruction
+  // will be changed to Nop and new folded instruction will be inserted before
+  // it.
+  bool ProcessOpSpecConstantOp(ir::Module::inst_iterator* pos);
+
+  // Try to fold the OpSpecConstantOp CompositeExtract instruction pointed by
+  // the given instruction iterator to a normal constant defining instruction.
+  // Returns the pointer to the new constant defining instruction if succeeded.
+  // Otherwise returns nullptr.
+  ir::Instruction* DoCompositeExtract(ir::Module::inst_iterator* inst_iter_ptr);
+
+  // Try to fold the OpSpecConstantOp VectorShuffle instruction pointed by the
+  // given instruction iterator to a normal constant defining instruction.
+  // Returns the pointer to the new constant defining instruction if succeeded.
+  // Otherwise return nullptr.
+  ir::Instruction* DoVectorShuffle(ir::Module::inst_iterator* inst_iter_ptr);
+
+  // Try to fold the OpSpecConstantOp <component wise operations> instruction
+  // pointed by the given instruction iterator to a normal constant defining
+  // instruction. Returns the pointer to the new constant defining instruction
+  // if succeeded, otherwise return nullptr.
+  ir::Instruction* DoComponentWiseOperation(
+      ir::Module::inst_iterator* inst_iter_ptr);
+
+  // Creates a constant defining instruction for the given Const instance and
+  // inserts the instruction at the position specified by the given instruction
+  // iterator. Returns a pointer to the created instruction if succeeded,
+  // otherwise returns a null pointer. The instruction iterator points to the
+  // same instruction before and after the insertion. This is the only method
+  // that actually manages id creation/assignment and instruction
+  // creation/insertion for a new Constant instance.
+  ir::Instruction* BuildInstructionAndAddToModule(
+      std::unique_ptr<analysis::Constant> c, ir::Module::inst_iterator* pos);
+
+  // Creates a Const instance to hold the constant value of the given
+  // instruction. If the given instruction defines a normal constants whose
+  // value is already known in the module, returns the unique pointer to the
+  // created Const instance. Otherwise does not create anything and returns a
+  // nullptr.
+  std::unique_ptr<analysis::Constant> CreateConstFromInst(
+      ir::Instruction* inst);
+
+  // Creates a Const instance with the given type and a vector of constant
+  // defining words. Returns an unique pointer to the create Const instance if
+  // the Const instance can be created successfully. Otherwise returns a null
```

Comment:
How can this fail, i.e. end up returning nullptr?  Seems it could only be in the composite constant case, were the operands are not themselves normal constants?